### ISSUE TITLE
Task 1: Create ML module structure and neural network core types

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ts-rs#47]
# Task 1: Create ML Module Structure and Neural Network Core Types

**Related to Epic**: #46

## Objective
Set up the foundational structure for the machine learning module with core neural network types, tensor operations, and GPU acceleration support.

## Scope
- Create `src/ml/mod.rs` with public API
- Define core neural network types and structures
- Implement tensor operations and GPU acceleration interfaces
- Set up model persistence and serialization

### File Structure
```
src/ml/
├── mod.rs              # Public API and re-exports
├── types.rs            # Core ML types and structures (new)
├── tensor.rs           # Tensor operations and GPU support (new)
├── activations.rs      # Activation functions (new)
└── persistence.rs      # Model serialization (new)
```

### Core Types to Implement
```rust
// Neural network architecture
pub struct NeuralNetwork {
    pub layers: Vec<Layer>,
    pub optimizer: OptimizerType,
    pub loss_function: LossFunction,
    pub training_config: TrainingConfig,
    pub device: Device,
}

// Layer definitions
pub enum Layer {
    Dense { input_size: usize, output_size: usize, activation: ActivationType },
    LSTM { input_size: usize, hidden_size: usize, num_layers: usize },
    GRU { input_size: usize, hidden_size: usize, num_layers: usize },
    Transformer { d_model: usize, n_heads: usize, d_ff: usize, n_layers: usize },
    Dropout { rate: f32 },
    BatchNorm { features: usize },
}

// Training configuration
pub struct TrainingConfig {
    pub learning_rate: f64,
    pub batch_size: usize,
    pub epochs: usize,
    pub validation_split: f64,
    pub early_stopping: Option<EarlyStoppingConfig>,
    pub device_preference: DevicePreference,
}

// Device abstraction for CPU/GPU
pub enum Device {
    CPU,
    CUDA(usize),  // GPU device index
    Metal,        // Apple Silicon
    Auto,         // Automatic selection
}
```

### 1. Tensor Operations
- **Basic Operations**: Matrix multiplication, addition, broadcasting
- **Gradient Computation**: Automatic differentiation support
- **Memory Management**: Efficient tensor allocation and deallocation
- **Device Management**: CPU/GPU tensor transfers

### 2. Activation Functions
- **Standard**: ReLU, Sigmoid, Tanh, Softmax
- **Advanced**: Swish, GELU, LeakyReLU, ELU
- **Custom**: Support for user-defined activation functions

### 3. GPU Acceleration Interface
- **CUDA Support**: GPU acceleration for NVIDIA GPUs
- **Metal Support**: Apple Silicon GPU acceleration
- **Fallback**: Graceful degradation to CPU when GPU unavailable
- **Memory Optimization**: Efficient GPU memory management

### 4. Model Persistence
- **Serialization**: Save/load trained models
- **Format Support**: Multiple formats (binary, JSON, ONNX compatibility)
- **Versioning**: Model version management and compatibility
- **Metadata**: Training history and hyperparameter storage

## Acceptance Criteria
- [ ] Module compiles without errors on all target platforms
- [ ] Core types support serialization with serde
- [ ] GPU acceleration works on available CUDA/Metal devices
- [ ] CPU fallback functions correctly when GPU unavailable
- [ ] Unit tests for all tensor operations and activations
- [ ] Documentation with architecture examples
- [ ] Integration with existing TimeSeries type

## Dependencies
- `candle-core` for tensor operations and GPU acceleration
- `candle-nn` for neural network layers and optimizers
- `serde` for model serialization
- `tch` (PyTorch bindings) as alternative backend option
- Existing `TimeSeries` and analysis modules

## Estimated Time
12-15 hours

## Next Steps
After completion, this enables:
- Task 2: LSTM and GRU implementation for sequential modeling
- Task 3: Transformer architecture for attention-based forecasting
- Task 4: Ensemble methods and model combination
